samples/web/content/apprtc/gcm_register.py

```diff
@@ -0,0 +1,204 @@
+#!/usr/bin/python2.4
+#
+# Copyright 2015 Google Inc. All Rights Reserved.
+
+"""AppRTC GCM ID Registration
+
+This module implements user ID registration and verification.
+"""
+
+import constants
+import datetime
+import logging
+import webapp2
+
+from google.appengine.ext import ndb
+
+def get_ancestor_key():
+  return ndb.Key('GCMRecord', 'global')
+
+class BindPage(webapp2.RequestHandler):
+  PARAM_USER_ID = 'userId'
+  PARAM_GCM_ID = 'gcmId'
+  PARAM_CODE = 'code'
+  PARAM_OLD_GCM_ID = 'oldGcmId'
+  PARAM_NEW_GCM_ID = 'newGcmId'
+
+  def get_message_from_json(self):
+    try:
+      message = json.loads(self.request.body)
+      return message
+    except Exception as e:
+      logging.warning('JSON load error from BindPage: error=' + str(e)
+          + ', request=' + self.request.body)
+      return None
+
+  def handle_new(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID]:
+      # TODO(jiayl): validate the input, generate a random code, and send SMS.
+      self.response.out.write(
+          GCMRecord.add_or_update(msg[PARAM_USER_ID], msg[PARAM_GCM_ID], ''))
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARG)
+
+  def handle_update(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] \
+        and msg[PARAM_OLD_GCM_ID] and msg[PARAM_NEW_GCM_ID]:
+      self.response.out.write(GCMRecord.update_gcm_id(
+          msg[PARAM_USER_ID], msg[PARAM_OLD_GCM_ID], msg[PARAM_NEW_GCM_ID]))
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARG)
+
+  def handle_verify(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID] and msg[PARAM_CODE]:
+      self.response.out.write(GCMRecord.verify(
+          msg[PARAM_USER_ID], msg[PARAM_GCM_ID], msg[PARAM_CODE]))
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARG)
+
+  def handle_del(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID]:
+      GCMRecord.remove(msg[PARAM_USER_ID], msg[PARAM_GCM_ID])
+      self.response.out.write(constants.RESPONSE_SUCCESS)
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARG)
+
+  def handle_query(self):
+    ids = self.get_message_from_json()
+    if ids:
+      result = []
+      for id in ids:
+        # TODO(jiayl): Only return the verified users when SMS verification is
+        # added.
+        if len(GCMRecord.get_by_user_id(id)) > 0:
+          result.append(id)
+      self.response.out.write(json.dumps(result))
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARG)
+
+  def get(self, cmd):
```

Comment:
I forgot to fix that before submit. Will fix.
On Jan 15, 2015 8:18 AM, "Chuck Hays" <notifications@github.com> wrote:

> In samples/web/content/apprtc/gcm_register.py
> <https://github.com/GoogleChrome/webrtc/pull/382#discussion-diff-23021096>
> :
>
> > +      self.response.out.write(constants.RESPONSE_INVALID_ARG)
> > +
> > +  def handle_query(self):
> > +    ids = self.get_message_from_json()
> > +    if ids:
> > +      result = []
> > +      for id in ids:
> > +        # TODO(jiayl): Only return the verified users when SMS verification is
> > +        # added.
> > +        if len(GCMRecord.get_by_user_id(id)) > 0:
> > +          result.append(id)
> > +      self.response.out.write(json.dumps(result))
> > +    else:
> > +      self.response.out.write(constants.RESPONSE_INVALID_ARG)
> > +
> > +  def get(self, cmd):
>
> Is get() in this framework mapped directly to GET requests? It seems like
> new/update/del should be POST/PUT/DELETE, or at least POST. We could also
> use standard http responses for many of the cases. Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleChrome/webrtc/pull/382/files#r23021096>.
>